CORENET-5972: Consume openvswitch-ipsec systemd service for OVN IPsec deployment

pperiyasamy opened the pull request:
This ipsec machine config extension is now installing `openvswitch3.5-ipsec` package on the node, so this PR consumes that package to configure, enable and start `openvswitch-ipsec` systemd service which basically moves away running ovs-monitor-ipsec process from container to host. 

It fixes following issues.

1. Getting rid of running host binaries inside ovn-ipsec container, for example `addconn` which depends on `glibc` which creates version incompatibility issues between host and container. 
2. There is no need for patching `auto=start` parameter into each IPsec connections which is introduced by the PR https://github.com/openshift/machine-config-operator/pull/4854.
3. The ovs-monitor-ipsec is now solving auth problem while using default crypto policies loaded by libreswan on the host.

In order to consume openvswitch-ipsec systemd service, this PR does the following:

1. Stop spawning ovs-monitor-ipsec as foreground process in the ovn-ipsec container. Instead setup required IPsec configuration parameters in the /etc/sysconfig/openvswitch file, enable and start the openvswitch-ipsec service
on the host. This is done at the time when ovn-ipsec-host pod is coming up for the first time. For the pod restart scenarios, it just checks openvswitch-ipsec service is running on the host, otherwise exit from the container with error.

2. Keep running an ovn-ipsec container that waits for the ovs-monitor-ipsec.log file to be created, monitors the openvswitch-ipsec service health, and redirects /var/log/openvswitch/ovs-monitor-ipsec.log to the ovn-ipsec container's stdout console. The container fails if the systemd service stops unexpectedly, ensuring
proper pod lifecycle management.

4. There is no necessity of doing ipsec state and policy cleanup in ovn-ipsec-cleanup container when OVN IPsec is handled via openvswitch-ipsec systemd service.

5. During the OCP upgrade, the new ipsec os extension takes a while to deploy with openvswitch3.5-ipsec package, so by the time ovn-ipsec-host daemonset is rendered, we need to handle that scenario by running ovs-monitor-ipsec in the container. So this PR is also considering the transition phase of the process that is moving
from container to host.

6. The ovn-keys init container configures ovs with IPsec certificate paths, so the container uses same host directory path to store and configure ovs with certificates because the ovs-monitor-ipsec process is running on the host now.

/assign @igsilya 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **Bug Fixes**
  * Improved IPsec certificate/key regeneration so the IPsec service restarts more reliably.
  * Enhanced coordination between the IPsec components in the pod and the host IPsec service for smoother startup and clean shutdown.

* **Refactor**
  * Updated IPsec asset storage and runtime configuration to use host-backed paths for more consistent behavior.
  * Refined volume mounts and command checks so health/liveness operations rely on the host IPsec tooling when available.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4854 https://github.com/openshift/os/pull/1718 https://github.com/openshift/machine-config-operator/pull/4878 https://github.com/openshift/ovn-kubernetes/pull/2472

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4854 https://github.com/openshift/os/pull/1718 https://github.com/openshift/machine-config-operator/pull/4878 https://github.com/openshift/ovn-kubernetes/pull/2472

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4854 https://github.com/openshift/os/pull/1718 https://github.com/openshift/machine-config-operator/pull/4878 https://github.com/openshift/ovn-kubernetes/pull/2472

pperiyasamy commented:
The `testwith` doesn't take https://github.com/openshift/os/pull/1718 changes, ipsec os extension is still not installing `openvswitch-ipsec` package, so we must get 1718 landed first for testing CNO changes.

igsilya commented:
@pperiyasamy I agree, we should get OVS 3.5 first into rhcos / ovn-k / microshift.  We can install openvswitch3.5-ipsec at the same time, it should not be a problem since the service is disabled until CNO activates it.  We need OVS 3.5 either way for other purposes (rhel 10 support, for example).   Once we have OVS 3.5 and the openvswitch-ipsec service we can more easily test CNO and other changes.

pperiyasamy committed:
Consume openvswitch-ipsec systemd service for OVN IPsec deployment

The consumption of openvswitch-ipsec systemd service from the host for the
ovn-ipsec-host deployment helps to avoid using any host binaries to be
running from ovn-ipsec container, so it solves protocol, glibc and hardware
compatibility issues when RHEL10 is supported for OCP platforms.

Hence this commit does the following:

1. Stop spawning ovs-monitor-ipsec as foreground process in the ovn-ipsec
container. Instead setup required IPsec configuration parameters in the
/etc/sysconfig/openvswitch file, enable and start the openvswitch-ipsec service
on the host. This is done at the time when ovn-ipsec-host pod is coming up
for the first time. For the pod restart scenarios, it just checks
openvswitch-ipsec service is running on the host, otherwise exit from the
container with error.

2. Keep running an ovn-ipsec container that waits for the ovs-monitor-ipsec.log
file to be created, monitors the openvswitch-ipsec service health, and redirects
/var/log/openvswitch/ovs-monitor-ipsec.log to the ovn-ipsec container's stdout
console. The container fails if the systemd service stops unexpectedly, ensuring
proper pod lifecycle management.

3. There is no necessity of doing ipsec state and policy cleanup in ovn-ipsec-cleanup
container when OVN IPsec is handled via openvswitch-ipsec systemd service.

4. During the OCP upgrade, the new ipsec os extension takes a while to deploy with
openvswitch3.5-ipsec package, so by the time ovn-ipsec-host daemonset is rendered,
we need to handle that scenario by running ovs-monitor-ipsec in the container.
So this commit is also considering the transition phase of the process that is moving
from container to host.

5. The ovn-keys init container configures ovs with IPsec certificate paths, so the
container uses same host directory path to store and configure ovs with certificates
because the ovs-monitor-ipsec process is running on the host now.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4854 https://github.com/openshift/os/pull/1718 https://github.com/openshift/machine-config-operator/pull/4878 https://github.com/openshift/ovn-kubernetes/pull/2472

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade openshift/os#1718 openshift/machine-config-operator#4878 openshift/ovn-kubernetes#2472

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade openshift/ovn-kubernetes#2472 openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/assign @anuragthehatter @huiran0826

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/ovn-kubernetes/pull/2472 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/assign @jcaamano

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

igsilya commented:
The change looks correct to me from the OVS and service ordering perspective, also tried it with cluster bot together with https://github.com/openshift/machine-config-operator/pull/4878 and it worked as expected (didn't test upgrades).
/lgtm
/retest

pperiyasamy commented:
/retest

pperiyasamy commented:
/reopen

pperiyasamy commented:
/remove-lifecycle rotten

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-ipsec-serial https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy committed:
Remove /usr/sbin host directory dependency from ovn-ipsec-host pod

This commit removes the mount of /usr/sbin from the host to eliminate
need for host binaries. it would make chroot to use container binary
which avoids GLIBC compatibility issue.

The 'ipsec --checknss' check is removed, it doesn't run during
upgrade until openvswitch3.5-ipsec is deployed to the node anyway,
so it's safe to skip during that transition period.

The liveness probe now uses 'ovs-appctl -t ovs-monitor-ipsec ipsec/status'
instead of 'ipsec whack --trafficstatus' to check for configured IPsec
tunnels without requiring host binaries.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy committed:
Remove /usr/libexec mount and avoid GLIBC compatibility issues

During OCP upgrade the new ovn-ipsec-host pod is rendered before the machine
config rolls out openvswitch3.5-ipsec package to node. This causes the
container to fall back to running ovs-monitor-ipsec in the container, but
mounting host's /usr/libexec shadows container binaries and causes GLIBC
compatibility errors (GLIBC_ABI_DT_RELR not found) when executing binaries
like addconn and platform-python.

- Remove /usr/libexec host mount entirely as it's no longer needed.
- Use container's /usr/libexec/platform-python binary to run ovs-monitor-ipsec
script.
- Remove addconn validation as /etc/ipsec.conf is managed by ovs-monitor-ipsec
and wait-for-ipsec-connect service, so it is always syntactically correct.
- Remove _stackmanager check as it's obsolete (removed in libreswan 5.3+) and OCP
must be using this version since 4.19.

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy committed:
Fix 'kill: not a pid' error by checking PID file before use

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/test ?

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22-e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

jluhrsen commented:
/testwith openshift/cluster-network-operator/master/5.0-upgrade-from-stable-4.22-e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
Two issues being noticed with e2e-aws-ovn-upgrade-ipsec run:
1. it uses build (which broke ipsec due to glibc incompatibility issue ) for the initial cluster, so ovn-ipsec-host pod is crashlooping until pod is upgraded to PR version.
2.  machine config daemon never deployed openvswitch3.5-ipsec package during upgrade and failing with below error.
```
2026-06-24T18:11:11.339388625Z W0624 18:11:11.339342   48949 update.go:1946] Extension package openvswitch3.5-ipsec not found in RPM database
2026-06-24T18:11:11.339444737Z E0624 18:11:11.339419   48949 writer.go:231] Marking Degraded due to: "extension package verification failed: the following extension packages are missing from the RPM database: [openvswitch3.5-ipsec]"
2026-06-24T18:11:11.392881358Z I0624 18:11:11.392825   48949 daemon.go:643] Error syncing node ip-10-0-16-239.ec2.internal (retries 2): extension package verification failed: the following extension packages are missing from the RPM database: [openvswitch3.5-ipsec]
2026-06-24T18:11:13.353789060Z I0624 18:11:13.353753   48949 daemon.go:796] Transitioned from state: Done -> Degraded

```
We had bug https://redhat.atlassian.net/browse/OCPBUGS-56555 for similar issue before, but looks like it's not fixed completely or don't know if it's a rhcos image issue for the ocp upgrade. will check with mco team.

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/release-5.0/5.0-upgrade-from-stable-4.22-e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

jluhrsen commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec openshift/machine-config-operator#4878

jluhrsen commented:
/testwith openshift/cluster-network-operator/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

jluhrsen commented:
/testwith openshift/cluster-network-operator/release-5.0/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

jluhrsen commented:
/testwith openshift/cluster-network-operator/master/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec openshift/ovn-kubernetes#3254 openshift/machine-config-operator#4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec openshift/ovn-kubernetes#3254 openshift/machine-config-operator#6250

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec openshift/machine-config-operator#6250

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/ovn-kubernetes/pull/3254 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/ovn-kubernetes/pull/3254 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -305,6 +364,10 @@ spec:
                  - |
                    #!/bin/bash
                    set -exuo pipefail
+                   if chroot /proc/1/root systemctl is-active --quiet openvswitch-ipsec; then
```
Why are our checks inconsistent? On some places we check if the service is defined, but here and in other places we check if the service is active? I might be missing something here.

pperiyasamy replied:
The service definition is checked only once in the ovn-ipsec container before configuring and starting `openswitch-ipsec` service (it happens only once), once service is started, it remains to active during the lifetime of the node. so all other places we only check if the service is active.

jcaamano replied:
so why not do `if grep -q "openshift.conf" /etc/sysconfig/openvswitch` here?

pperiyasamy replied:
done.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,56 @@ spec:
+                echo "$ovsipsecservice service stopped unexpectedly"
+                exit 1
+              fi
+              tail -F /var/log/openvswitch/ovs-monitor-ipsec.log
```
I think `tail -F` is pretty resilient and wont fail if ipsec service stops. Have you tested it?

pperiyasamy replied:
you are right, `tail -F` doesn't fail upon `openvswitch-ipsec` service stop, so service crash would never get unnoticed. now checking service activeness with subshell, going to test this change with a cluster.

pperiyasamy replied:
It's the good suggestion,thanks.  as we discussed, changed it with  `journalctl -f _SYSTEMD_INVOCATION_ID="$(systemctl show -p InvocationID --value openvswitch-ipsec)" -n all` to show only logs since last service restart.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,56 @@ spec:
+              if ! chroot /proc/1/root systemctl is-active --quiet $ovsipsecservice; then
+                echo "$ovsipsecservice service stopped unexpectedly"
+                exit 1
+              fi
```
Wouldn't it be simpler to just have a subshell periodically checking if the service is active and sending a SIGTERM to the parent shell when not?

Also, I am not sure how tight you want to align the state of the service to the state of the pod. The service will restart by itself and you might miss it.

pperiyasamy replied:
we run `ovs-monitor-ipsec` as a ovn-ipsec container foreground process, container will be restarted when script fails, so wanted to reflect the same for `openvswitch-ipsec` systemd service. hope that should be ok.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,56 @@ spec:
 
+          # Configure IPsec using the host's openvswitch-ipsec systemd service if present.
+          # This avoids running host binaries from the container and solves compatibility issues.
+          # Fall back to container-based ovs-monitor-ipsec for older nodes without the service.
```
Why keep the fallback? Is not working in all circumstances anyway, is it?

jcaamano replied:
Maybe to keep the upgrade working , can remove after?

pperiyasamy replied:
that's right, this `ovs-monitor-ipsec` fallback is needed at the time of upgrade until node is upgraded to use `openvswitch-ipsec` service. updated the comment.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -440,7 +501,15 @@ spec:
+            # It means ovs-monitor-ipsec process is running on the host always, so ip
+            # xfrm state and policy entries are managed properly according to ipsec
+            # configuration on the OVN. so skip flushing ipsec for this case.
+            if chroot /proc/1/root systemctl is-active --quiet openvswitch-ipsec; then
```
```
When IPsec connections are managed by openvswitch-ipsec systemd service
```
When is this? When `openvswitch-ipsec` is running, or when it is present? It should be a static configuration, so it should be when it is present even if it is not running which would indicate a different circumstance?

pperiyasamy replied:
The `openvswitch-ipsec` service is just a replacement for running `ovs-monitor-ipsec` process from container and it's managing entries only with `openshift.conf` file, pluto is still managing ip xfrm state/policy entries. updated the comment to reflect this, is that clear now ?

jcaamano replied:
Again, why not check `grep -q "openshift.conf" /etc/sysconfig/openvswitch` here?

pperiyasamy replied:
sure, done

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -338,13 +338,6 @@ spec:
           # Workaround for https://github.com/libreswan/libreswan/issues/373
           ulimit -n 1024
 
-          /usr/libexec/ipsec/addconn --config /etc/ipsec.conf --checkconfig
-          # Check kernel modules only for libreswan version <= 5.2. The _stackmanager binary is
-          # removed from 5.3 onwards, so this check is not needed on later versions.
-          if [ -e /usr/libexec/ipsec/_stackmanager ]; then
-            /usr/libexec/ipsec/_stackmanager start
-          fi
-
```
So I don't understand a couple of things:

Commit message says
```
- Use container's /usr/libexec/platform-python binary to run ovs-monitor-ipsec
script.
```
But I don't see such change.

Also if we manage to keep the fallback working with commits `Remove /usr/sbin host directory dependency from ovn-ipsec-host pod` and `Remove /usr/libexec mount and avoid GLIBC compatibility issues` I don't understand why we need anything else.

pperiyasamy replied:
> - Use container's /usr/libexec/platform-python binary to run ovs-monitor-ipsec
script.
But I don't see such change.

I just meant existing script with the above. since /usr/libexec from host is not mounted now, it uses container's  `/usr/libexec/platform-python` binary to run `ovs-monitor-ipsec` script.

```
          # Start ovs-monitor-ipsec which will monitor for changes in the ovs
          # tunnelling configuration (for example addition of a node) and configures
          # libreswan appropriately.
          # We are running this in the foreground so that the container will be restarted when ovs-monitor-ipsec fails.
          /usr/libexec/platform-python /usr/share/openvswitch/scripts/ovs-monitor-ipsec \
            --pidfile=/var/run/openvswitch/ovs-monitor-ipsec.pid --ike-daemon=libreswan --no-restart-ike-daemon \
            --ipsec-conf /etc/ipsec.d/openshift.conf --ipsec-d /var/lib/ipsec/nss \
            --log-file --monitor unix:/var/run/openvswitch/db.sock
```
https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/ovn-kubernetes/common/ipsec-host.yaml#L291-L298

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -472,14 +541,20 @@ spec:
+          # When ovs-monitor-ipsec process is not running via openvswitch-ipsec systemd
+          # service, then keep monitoring ovn-ipsec container process until it terminates
+          # in the ovn-ipsec container's pre stop hook.
+          if ! chroot /proc/1/root systemctl is-active --quiet openvswitch-ipsec; then
```
why not check grep -q "openshift.conf" /etc/sysconfig/openvswitch here?

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,66 @@ spec:
+              fi
+              sleep 1
+            done
+
```
I don't think we need any of this?

pperiyasamy replied:
removed unwanted checks now, but still waiting for `ovs-monitor-ipsec.pid` to be created and can copied to shared file for liveness probe.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,66 @@ spec:
+              done
+            ) &
+            watcher_pid=$!
+            trap 'kill $watcher_pid 2>/dev/null; exit 1' TERM
```
Why do you need this?

pperiyasamy replied:
removed it, as per your suggestion, doing service activeness check via liveness probe.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,66 @@ spec:
+              set +x
+              while true; do
+                sleep 5
+                if ! chroot /proc/1/root systemctl is-active --quiet $ovsipsecservice; then
```
if you want to tightly align the lifecycle of the pod with the lifecycle of the service, you might want to actually monitor PID changes

pperiyasamy replied:
liveness probe uses PID changes now.

`Warning  Unhealthy  8s                 kubelet            Liveness probe failed: openvswitch-ipsec service restarted (PID changed from 25992 to 35600)`

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,66 @@ spec:
+            watcher_pid=$!
+            trap 'kill $watcher_pid 2>/dev/null; exit 1' TERM
+            while true; do
+              tail -F /var/log/openvswitch/ovs-monitor-ipsec.log
```
the problem here is that if the service exists and restarts, and the the pod exists and restarts, some of the logs of the new service run will not be on the new pod

you might want to ensure when tailing that you get all the contents of the file instead of just the last 10 lines with `tail -n +1`
But that might also include past service runs. Do you know if the service restart purges the log file?

pperiyasamy replied:
openvswitch-ipsec service restart doesn't purge the log file, but we are good now with new journalctl command.

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/ovn-kubernetes/pull/3254 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/pipeline required

pperiyasamy commented:
/testwith openshift/ovn-kubernetes/main/5.0-upgrade-from-stable-4.22/e2e-aws-ovn-upgrade-ipsec https://github.com/openshift/ovn-kubernetes/pull/3254 https://github.com/openshift/machine-config-operator/pull/4878

pperiyasamy commented:
/testwith openshift/cluster-network-operator/master/e2e-aws-ovn-serial-ipsec https://github.com/openshift/machine-config-operator/pull/4878

igsilya commented:
@pperiyasamy I didn't test this code and I'm not an expert in pod lifecycles, but the change seems correct to me.  There is a ton of fallback code for the old behavior, which I hope can be removed in 5.1 (users can't upgrade to 5.1 without upgrading to 5.0 first, right?).  Otherwise,
/lgtm

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,41 @@ spec:
+          # Fall back to container-based ovs-monitor-ipsec when the openvswitch-ipsec systemd service
+          # is not yet available on the node.
+          ovsipsecservice="openvswitch-ipsec"
+          if chroot /proc/1/root systemctl list-unit-files --type=service | grep "$ovsipsecservice"; then
```
use `grep -q`

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,41 @@ spec:
+                  exit 1
+                fi
+                sleep 1
+              done
```
I don't think we need to check this, if the service is enabled, the pid needs to be there. And if it is not, we can just bail out with the suggestion in my next comment

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -249,6 +258,41 @@ spec:
+            fi
+
+            # Store initial PID for liveness probe to detect service restarts
+            cp /var/run/openvswitch/ovs-monitor-ipsec.pid /tmp/ovs-monitor-ipsec-base.pid
```
I guess I would do something like this:
```
              # Capture PID and InvocationID atomically, verify no restart occurred
              invocation_id=$(chroot /proc/1/root systemctl show -p InvocationID --value openvswitch-ipsec)
              main_pid=$(chroot /proc/1/root systemctl show -p MainPID --value openvswitch-ipsec)
              file_pid=$(cat /var/run/openvswitch/ovs-monitor-ipsec.pid 2>/dev/null)
              if [ "$main_pid" != "$file_pid" ]; then
                echo "openvswitch-ipsec PID mismatch: systemd reports $main_pid, PID file has $file_pid"
                exit 1
              fi
              echo "$main_pid" > /tmp/ovs-monitor-ipsec-base.pid

              while true; do
                chroot /proc/1/root journalctl -f "_SYSTEMD_INVOCATION_ID=$invocation_id" -n all
                echo "journalctl exited unexpectedly, retrying..."
                sleep 2
              done
```

So we capture invocation_id and pid right away, to frame what service run we are monitoring.

pperiyasamy replied:
nice, updated it.

jcaamano commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -360,9 +390,26 @@ spec:
+                  file_pid=$(cat /var/run/openvswitch/ovs-monitor-ipsec.pid 2>/dev/null)
+                  main_pid=$(cat /tmp/ovs-monitor-ipsec-main.pid)
+                  if [ "$main_pid" != "$file_pid" ]; then
+                    echo "openvswitch-ipsec PID mismatch: systemd reports $main_pid, PID file has $file_pid"
```
I would keep the original message you had here

jcaamano reviewed (commented):
small nit, otherwise lgtm

igsilya commented on bindata/network/ovn-kubernetes/common/ipsec-host.yaml:
```diff
@@ -332,16 +368,10 @@ spec:
           name: host-var-lib
         - mountPath: /etc
           name: host-etc
-        - mountPath: /usr/sbin
-          name: usr-sbin
-        - mountPath: /usr/libexec
-          name: usr-libexec
```
Hmm.  The code still supports both modes - systemd on the host or `ovs-monitor-ipsec` daemon inside the container.  But if we do not mount these, then `ovs-monitor-ipsec` inside the container will use `ipsec` command from the inside of the container that is potentially incompatible with the `pluto` running on the host.  Is that right?  Should the mounts be preserved until the fallback code is removed?

jcaamano replied:
I guess this comment is inline with the comment [here](https://github.com/openshift/cluster-network-operator/pull/2662#discussion_r3579438607) 
```
Also if we manage to keep the fallback working with commits Remove /usr/sbin host directory dependency from ovn-ipsec-host pod and Remove /usr/libexec mount and avoid GLIBC compatibility issues I don't understand why we need anything else.
```

I talked with @pperiyasamy offline and what I understood is this is picking between two evils: either we live RHEL10 glibc problems or we live with our host/container protocol problems and he chose the latter as the most unlikely. Please @pperiyasamy confirm.

pperiyasamy replied:
yes, mounting `/usr/sbin` also broke with `chroot` command (which pod uses it heavily) due to glibc compatibility issues. so had to get rid of mounting host `/usr/sbin` directory completely.

jcaamano replied:
And I guess this would be the only thing that would work in an hypothetical RHEL9 -> RHEL10 upgrade case, which I am not sure its a thing.

igsilya replied:
Yeah, I see, it's hard to actually have both cases 100% working without doing terrible things inside the container like building a separate chroot or something.  Since RHEL10 is TP in 5.0, I suppose, sacrificing the 4.x RHEL 9 -> 5.0 RHEL 10 upgrade path is more reasonable than having breakages in common cases.  May need some documentation around that, but otherwise sounds fine to me.

jcaamano commented:
/lgtm
/approve

jcaamano commented:
/retest

mkowalski commented:
/retest-required

mkowalski commented:
/verified by CI

mkowalski commented:
/override-sticky ci/prow/e2e-aws-ovn-upgrade-ipsec
/override-sticky ci/prow/e2e-aws-ovn-fdp-qe
/override-sticky ci/prow/e2e-metal-ipi-ovn-dualstack-bgp-local-gw
